lib/db/schema/patch-029-30.sql

```diff
@@ -0,0 +1,32 @@
+-- Update deleteSessionToken stored procedure to delete from devices.
+CREATE PROCEDURE `deleteSessionToken_3` (
+  IN `tokenIdArg` BINARY(32)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE tokenId = tokenIdArg;
+  DELETE FROM unverifiedTokens WHERE tokenId = tokenIdArg;
+  DELETE FROM devices WHERE sessionTokenId = tokenIdArg;
+
+  COMMIT;
+END;
+
+-- Prior to this migration, it was possible to delete a session token
+-- without deleting the associated device (if a device record exists).
+-- That's fixed by deleteSessionTokens_3, but we should also purge any
+-- historical devices that have reached this state.
+DELETE FROM devices
+  USING devices LEFT JOIN sessionTokens
+    ON devices.sessionTokenId = sessionTokens.tokenId
+  WHERE sessionTokens.tokenId IS NULL;
```

Comment:
We should also consider the performance implications of doing this against large tables in production.  I wonder if it will cause a window of write outage on the sessionstokens and/or devices table.